### PR TITLE
Add PrivateArray kernel_api

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Creates `PrivateArray` class that is used inside `kernel_api`. Overloaded with `USMNdArray` instance generation on stack.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/kernel_api_impl/spirv/arrayobj.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    ty_array: types.Array,
+    ty_shape: Union[types.IntegerLiteral, types.BaseTuple],
+    shape: llvmir.Value,
+    data: llvmir.Value,
+):
+    """Makes SPIR-V array and fills it data."""
+    # Create array object
+    ary = context.make_array(ty_array)(context, builder)
+
+    itemsize = get_itemsize(context, ty_array)
```
Comment (diptorupd) [inline numba_dpex/kernel_api_impl/spirv/arrayobj.py:55]: Did you verify that we do not need the LLVM codegen target for itemsize here?

**Review by diptorupd — CHANGES_REQUESTED** [review]
Few technical questions. The docstrings all have copypasta issues.

File: numba_dpex/kernel_api_impl/spirv/arrayobj.py (hunk 1 above)
Comment (ZzEeKkAa) [reply numba_dpex/kernel_api_impl/spirv/arrayobj.py:55]: I'm using the same target as `numba_dpex.private.array`. See the implementation:
```python
def get_itemsize(context: SPIRVTargetContext, array_type: types.Array):
    """
    Return the item size for the given array or buffer type.
    Same as numba.np.arrayobj.get_itemsize, but using spirv data.
    """
    targetdata = _get_target_data(context)
    lldtype = context.get_data_type(array_type.dtype)
    return lldtype.get_abi_size(targetdata)
```

File: numba_dpex/kernel_api_impl/spirv/arrayobj.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    context: BaseContext,
+    builder: IRBuilder,
+    ty_array: types.Array,
+    ty_shape: Union[types.IntegerLiteral, types.BaseTuple],
+):
+    """Allocates flat array of given shape on the stack."""
+    if not isinstance(ty_shape, types.BaseTuple):
+        ty_shape = (ty_shape,)
+
+    return cgutils.alloca_once(
```
[the remark this answers is not in the document]
Comment (ZzEeKkAa) [reply numba_dpex/kernel_api_impl/spirv/arrayobj.py:101]: Yes,
```python
def alloca_once(builder, ty, size=None, name='', zfill=False):
    """Allocate stack memory at the entry block of the current function
    pointed by ``builder`` with llvm type ``ty``.  The optional ``size`` arg
    set the number of element to allocate.  The default is 1.  The optional
    ``name`` arg set the symbol name inside the llvm IR for debugging.
    If ``zfill`` is set, fill the memory with zeros at the current
    use-site location.  Note that the memory is always zero-filled after the
    ``alloca`` at init-site (the entry block).
    """
```

**Review by diptorupd — APPROVED** [review]
Thank you!

**Pull request merged**